### PR TITLE
Make Predictive a PyroModule

### LINKED ISSUE #2154: Refactor Predictive to make it a serializable PyroModule
While working on #2083, I found that the Predictive class is not working for AutoGuides as expected. 

One issue is that using a Predictive module where we set model and trained guide as submodules throws an error `submodule model has executed outside of supermodule".format(name)`.

It will be nice to refactor it so that it:
 - [ ] Returns correct results for autoguides.
 - [ ] Supports serializability as `ScriptModule`.
 - [ ] Faster serializability for the sequential predictive function.

### PR DESCRIPTION
**Pull request by neerajprad**
Attempts to fix #2154 - added a failing test.

### CODE REVIEW

File: pyro/infer/predictive.py [hunk 1]
```diff
@@ -163,13 +180,15 @@ def get_samples(self, *args, **kwargs):
         """
         posterior_samples = self.posterior_samples
         if self.guide is not None:
-            # use return_sites='' as a special signal to return all sites
```
Comment (fehiepsi) [inline pyro/infer/predictive.py]: I wonder if docs for `return_sites`: "by default only sample sites not present in `posterior_samples` are returned" is still valid now? The old behavior makes more sense to me.

Comment (neerajprad) [reply pyro/infer/predictive.py]: You mean for the case where we have both a guide and non-empty `posterior_samples`?

Comment (neerajprad) [reply pyro/infer/predictive.py]: I was thinking of actually raising an exception in that case, and adding support for it later if requested.

Comment (fehiepsi) [reply pyro/infer/predictive.py]: I mean for the case `return_sites=None`, we will include all sites by default with this change. The former behavior is "only sample sites not present in posterior_samples are returned".

File: pyro/infer/predictive.py [hunk 2]
```diff
@@ -66,15 +67,15 @@ def _predictive(model, posterior_samples, num_samples, return_sites=None,
...
         append_ndim = max_plate_nesting - len(model_trace.nodes[site]["fn"].batch_shape)
         site_shape = (num_samples,) + (1,) * append_ndim + model_trace.nodes[site]['value'].shape
-        if isinstance(return_sites, (list, tuple, set)):
+        if return_sites is not None:
             if site in return_sites:
                 return_site_shapes[site] = site_shape
         else:
-            if (return_sites is not None) or (site not in reshaped_samples):
-                return_site_shapes[site] = site_shape
+            # include all sites by default
```
Comment (fehiepsi) [inline pyro/infer/predictive.py]: @neerajprad I meant this case.

Comment (neerajprad) [reply pyro/infer/predictive.py]: That's a mistake, thanks for catching! Can I just use the `if site not in posterior_samples` check here, or do we really need the sentinel `''` value?

File: pyro/infer/predictive.py [hunk 3]
```diff
@@ -99,7 +100,7 @@ def _predictive(model, posterior_samples, num_samples, return_sites=None,
     return predictions
 
 
-class Predictive:
+class Predictive(PyroModule):
```
Comment (fritzo) [inline pyro/infer/predictive.py]: Hmm, this will currently break in a few ways:
1. in `PyroModule._pyro_set_supermodule()`, `assert not value._pyro_context.used` will fail
2. names will change in `model` and `guide`, invalidating the `PyroModule`-paramstore coupling.

Maybe we should make `Predictive` a `torch.nn.Module` instead?

Comment (neerajprad) [reply pyro/infer/predictive.py]: Yes, I bumped into exactly those two issues while debugging. 😞 Let me try your suggestion instead.

File: pyro/infer/predictive.py (hunk 2 above)
Comment (fehiepsi) [reply pyro/infer/predictive.py]: I think using `''` is not needed if we raise an exception where both `guide` and `posterior_samples` are specified. I don't know in which situation users need both of them. :D

File: pyro/infer/predictive.py (hunk 3 above)
Comment (fritzo) [reply pyro/infer/predictive.py]: On further thought, I don't see how `PyroModule` is appropriate. `nn.Module` seems like the right way to serialize. But I don't fully understand the issues...

Comment (fritzo) [reply pyro/infer/predictive.py]: For example if `model` and `guide` have `PyroSample` sites, those addresses must match exactly, and that can't happen in a single `PyroModule`, but is fine if they live in a common `nn.Module`.

Comment (neerajprad) [reply pyro/infer/predictive.py]: You are right - the poutine calls should also be serialized correctly. I am chasing some unrelated issues though, still debugging.

File: pyro/infer/predictive.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/predictive.py]: I think you original cases are indeed needed. :) Seems like we would need to distinguish between three cases, but I'm just using a `()` instead of `''`.
 - when `return_sites` is non empty we return samples from it.
 - when `return_sites = ()` (default), we return samples not in posterior samples.
 - when `return_sites = None` and a `guide` argument is provided (only used internally when a guide is provided), we return all sites. This is important because a guide will pass in `posterior_samples` but we should return those samples to the user.

Comment (fehiepsi) [reply pyro/infer/predictive.py]: Yeah, that looks good to me!

File: tests/infer/test_autoguide.py [hunk 4]
```diff
@@ -596,22 +597,77 @@ def test_linear_regression_smoke(auto_class, Elbo):
...
+        warnings.filterwarnings("ignore", category=torch.jit.TracerWarning)
+        traced_predictive = torch.jit.trace_module(predictive, {"call": (x,)})
+    f = io.BytesIO()
+    torch.jit.save(traced_predictive, f)
+    f.seek(0)
+    predictive_deser = torch.jit.load(f)
+    pyro.set_rng_seed(0)
+    samples_deser = predictive_deser.call(x)
+    # Note that the site values are different in the serialized guide
+    assert len(samples) == len(samples_deser)
```
Comment (neerajprad) [inline tests/infer/test_autoguide.py]: This works but I the traced predictive module does not give exactly the same values as the original one (fixing the `rng` seed). I think this should be good for now, but I'll try to harden the test if I figure out why the traced module gives different results.

**Comment by fritzo** [comment]
LGTM on the `nn.Module` refactoring. @fehiepsi have you reviewed the other changes?

File: pyro/nn/module.py [hunk 5]
```diff
@@ -161,8 +164,11 @@ def _pyro_set_supermodule(self, name, context):
         self._pyro_context = context
         for key, value in self._modules.items():
             assert isinstance(value, PyroModule)
-            assert not value._pyro_context.used, \
+            assert not value._pyro_context.active, \
                 "submodule {} has executed outside of supermodule".format(name)
+            if value._pyro_context.used:
```
Comment (fritzo) [inline pyro/nn/module.py]: Why is this needed?

Comment (neerajprad) [reply pyro/nn/module.py]: Hmm..sorry, none of this is needed. This was from a debugging session that I thought was reverted in the PR. Will remove.
